File: impl/app/android/app/src/main/java/com/ak/contexta/domain/usecase/GetHomeArticlesUseCase.kt
package com.ak.contexta.domain.usecase

import com.ak.contexta.domain.generation.categoryToDifficulty
import com.ak.contexta.domain.model.Article
import com.ak.contexta.domain.model.ArticleStatus
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 从 HomeViewModel 提取的文章过滤和排序逻辑。
 * 领域规则：按用户难度匹配文章，不足时返回全部已生成文章。
 */
@Singleton
class GetHomeArticlesUseCase @Inject constructor() {

    /**
     * 过滤并排序当前批次的文章。
     *
     * @param articles 批次所有文章
     * @param userDifficulty 用户当前难度等级
     * @param displayLimit 每批展示数量
     * @return 过滤排序后的文章列表
     */
    operator fun invoke(
        articles: List<Article>,
        userDifficulty: String,
        displayLimit: Int
    ): List<Article> {
        val matching = articles
            .filter { it.status != ArticleStatus.PENDING }
            .filter { categoryToDifficulty(it.contentCategory) == userDifficulty }
            .sortedBy { it.orderIndex }
            .take(displayLimit)

        return if (matching.isNotEmpty()) {
            matching
        } else {
            articles
                .filter { it.status != ArticleStatus.PENDING }
                .sortedBy { it.orderIndex }
                .take(displayLimit)
        }
    }
}

File: impl/app/flutter/lib/domain/usecase/get_home_articles_usecase.dart
import '../generation/article_prompts.dart';
import '../model/article.dart';

/// 从 HomeViewModel 提取的文章过滤和排序逻辑（对照 Kotlin GetHomeArticlesUseCase.kt）。
///
/// ⚠️ [displayLimit] 来自 `daily_learning.dailyCountSnapshot`（学习分配时抓拍的用户设置），
/// 不是当前 `user_settings.daily_article_count`。修改篇数设置后，已有阅读记录不受影响。
///
/// 领域规则：按用户难度匹配文章，不足时返回全部已生成文章。
class GetHomeArticlesUseCase {
  /// 过滤并排序当前批次的文章。
  ///
  /// [articles] 批次所有文章（同步批次每批 5 篇；2026-08-13 T6 生成管道
  /// 删除后无固定上限，按服务端实际返回）
  /// [userDifficulty] 用户当前难度等级（取自 `user_settings`）
  /// [displayLimit] 每批展示数量（取自 `daily_learning.dailyCountSnapshot`，可能 ≠ 当前设置）
  List<Article> call(
    List<Article> articles,
    String userDifficulty,
    int displayLimit,
  ) {
    final matching = articles
        .where((a) => a.status != ArticleStatus.pending)
        .where((a) => categoryToDifficulty(a.contentCategory) == userDifficulty)
        .toList()
      ..sort((a, b) => a.orderIndex.compareTo(b.orderIndex));

    final candidates = matching.isNotEmpty
        ? matching
        : articles.where((a) => a.status != ArticleStatus.pending).toList()
          ..sort((a, b) => a.orderIndex.compareTo(b.orderIndex));

    return candidates.take(displayLimit).toList();
  }
}
